Piggy back on request promise for 2.5 seconds

wwelling opened the pull request:
# Description

An optimization workaround that prevents weaver applications from making duplicate requests from models and repos. This is common when directives and controllers are not carefully scoped.

There is a hardcoded 2500 millisecond timeout for duplicate requests to share (piggy back) a request promise made before. This may be best to be in appConfig.js.

## Type of change

- [x] Optimization workaround

# How Has This Been Tested?

Please check box below when reviewed.

- [x] [Vireo](https://github.com/TexasDigitalLibrary/Vireo)

tamu-sad-iii committed:
Piggy back on request promise for 2.5 seconds

kaladay commented on app/model/abstractModel.js:
```diff
@@ -59,9 +61,20 @@ core.factory("AbstractModel", function ($injector, $rootScope, $q, ModelCache, M
                 if (fetch) {
-                    WsApi.fetch(mapping.instantiate).then(function (res) {
+                    if (fetchingPromise === undefined) {
+                        console.log('fetching', mapping.instantiate);
```
Could we avoid console logging on every fetch?
I am referring to this console log ferenced and the others added.

kaladay commented on app/repo/abstractRepo.js:
```diff
@@ -383,10 +385,23 @@ core.service("AbstractRepo", function ($q, $rootScope, $timeout, ApiResponseActi
+                if (fetchingAllPromise === undefined) {
+                    console.log('fetching', abstractRepo.mapping.all);
+                    fetchingAllPromise = WsApi.fetch(abstractRepo.mapping.all);
+                } else  {
```
Please remove extra space after `else`.

tamu-sad-iii committed:
Removed console logs

kaladay reviewed (commented):
What happens when two or more requests are handling the same request but one fails before the other completes?

It looks like the promise is being removed and a reject is being returned.

Could the rejection affect the other request?
Has this been tested?

ghost commented:
> What happens when two or more requests are handling the same request but one fails before the other completes?

_The first request fail/pass state will determine response for duplicates within 2500 milliseconds._

> It looks like the promise is being removed and a reject is being returned.

_The error is being caught and rejected appropriately from the Web Service API. And effectively removing the definition from the promise._

> Could the rejection affect the other request? Has this been tested?

_For the duration of the defer, the promise will be handled if subscribed to already. Yes, it has been tested in the Vireo admin submission list view._

kaladay commented:
> The subsequent requests will have a failed state if a duplicate request is issued within approximately 2.5 seconds.

This is a huge concern of mine.

Moving on, analytics shows some odd behavior when doing some testing.

Visually speaking when using the changes here the page appears to load faster.
When viewing the network log the traffic appears to block less.

When using the performance tester and even Lighthouse, the weaver changes appears to take longer and be slower. Looking at the data I feel inclined to consider Lighthouse unreliable.

Note: These tests are against the Vireo branch without the associated Vireo PR and without these changes verses the branch with the associated Vireo PR and these changes. Some of the differences observed here may be related to the Many to Many Lazy change in Vireo.

wwelling commented:
Can you perform each test 5 times and provide the average as well?

kaladay commented:
> Can you perform each test 5 times and provide the average as well?

No, you can go ahead and do that if you want to.


For better analysis, here are the analysis comparing with/without vireo changes.

### XHR Traffic With These Changes but Without Vireo Changes
![with_weaver_changes-without_vireo_changes-xhr_traffic](https://user-images.githubusercontent.com/35114839/234411770-6d0d2730-ec0d-467c-8d0e-8fe480012f62.png)

### XHR Traffic Without These Changes but With Vireo Changes
![without_weaver_changes-with_vireo_changes-xhr_traffic](https://user-images.githubusercontent.com/35114839/234411853-b9c120eb-4458-46d1-8f66-560bf1b4757c.png)

### Performance With These Changes but Without Vireo Changes
![with_weaver_changes-without_vireo_changes-performance](https://user-images.githubusercontent.com/35114839/234411928-6dac09ed-f652-47ad-92d2-00b9769ffb73.png)

### Performance Without These Changes but With Vireo Changes
![without_weaver_changes-with_vireo_changes-performance](https://user-images.githubusercontent.com/35114839/234412090-293facde-ec6d-4ebe-9d77-4438d74b96ef.png)

### Lighthouse With These Changes but Without Vireo Changes
![with_weaver_changes-without_vireo_changes-lighthouse](https://user-images.githubusercontent.com/35114839/234412142-06b2d12d-d246-4b6f-9b4c-efa3392472ed.png)

### Lighthouse Without These Changes but With Vireo Changes
![without_weaver_changes-with_vireo_changes-lighthouse](https://user-images.githubusercontent.com/35114839/234412201-91cb7bb4-1483-4d58-a9cd-a66caf241be0.png)

wwelling commented:
Please remove all local machine measurements from the review comments please.

wwelling commented:
> Some of the differences observed here may be related to the Many to Many Lazy change in Vireo.

Can you perform your local measured tests with or without the @wvr/core dependency change?

ghost commented:
> This is a huge concern of mine.

What is the concern? Duplicate requests are not appropriate by any means by a client application. We are not long polling; we have adopted web socket communications.

wwelling commented:
> > The subsequent requests will have a failed state if a duplicate request is issued within approximately 2.5 seconds.
> 
> This is a huge concern of mine.

Is the concern my poor grammar? I must have to read and write more.

tamu-sad-iii committed:
Merge branch '2.2.5-rc.3' into piggy-back-poorly-scoped-requests

ghost commented:
@kaladay I am willing to put a configuration in appConfig.js to enable/disable and another for the wait for duplicate requests. This will afford us to disable in Vireo and fix the scoping problems that are causing the duplicate requests. Other weaver ui apps are far less likely to be making duplicate requests.

ghost commented:
@jeremythuff @jcreel is this safe to merge and cut a rc?

tamu-sad-iii committed:
Timeout allowing control duplicate requests

Usage in Vireo

```
Vireo/src/main/webapp/app/config/apiMapping.js
```

https://github.com/TexasDigitalLibrary/Vireo/commit/dd0cf009d55716bfd3aeac4087e83dada30f6fff

tamu-sad-iii committed:
Better conditions
